### PR TITLE
Fix a trivial issue in how IsLifted is calculated for C# compound ass…

### LINKED ISSUE #21723: Address comments about how IsLifted is calculated for operators
See https://github.com/dotnet/roslyn/pull/21407#discussion_r135119021.

Please also add relevant unit tests.

### PR DESCRIPTION
**Pull request by mavasani**
…ignment operator

Fixes https://github.com/dotnet/roslyn/issues/21723

NOTE: We already have unit tests covering it. This just fixes the underlying implemented used to compute the value of this flag.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 1]
```diff
@@ -856,7 +856,7 @@ private ICompoundAssignmentExpression CreateBoundCompoundAssignmentOperatorOpera
             BinaryOperatorKind operatorKind = Helper.DeriveBinaryOperatorKind(boundCompoundAssignmentOperator.Operator.Kind);
             Lazy<IOperation> target = new Lazy<IOperation>(() => Create(boundCompoundAssignmentOperator.Left));
             Lazy<IOperation> value = new Lazy<IOperation>(() => Create(boundCompoundAssignmentOperator.Right));
-            bool isLifted = boundCompoundAssignmentOperator.Type.IsNullableType();
+            bool isLifted = boundCompoundAssignmentOperator.Operator.Kind.IsLifted();
```
Comment (mavasani) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: This seems to be correctly computed using `OperatorKind.IsLifted()` at other places in this file, only this instance was different.

**Review by jinujoseph — APPROVED** [review]

**Comment by mavasani** [comment]
@dotnet-bot retest microbuild_prtest  please

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: 👍

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Should this change affect a test output?

Comment (mavasani) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: I don't think so - this was just an observation from @AlekseyTs during a previous PR.

**Comment by AlekseyTs** [comment]
> Fixes #21407

It looks like #21407 is a PR rather than issue. What issue does this PR fix?

**Comment by mavasani** [comment]
@AlekseyTs Sorry, the issue being fixed is https://github.com/dotnet/roslyn/issues/21723

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Seems like a test gap. Could you add one that checks this property?

**Comment by AlekseyTs** [comment]
Please add a test that would fail without this fix.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1).

**Comment by mavasani** [comment]
@jcouv @AlekseyTs  Added unit test - verified it fails prior to the fix.

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 3)

**Review by jcouv — APPROVED** [review]
LGTM Thanks

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_unit64_prtest please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_release_unit64_prtest please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_release_vs-integration_prtest please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_unit64_prtest please

**Comment by mavasani** [comment]
Test failed due to flaky test https://github.com/dotnet/roslyn/issues/14592

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_unit64_prtest please

File: src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```
Comment (TyOverby) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs]: What is changed here?

**Review by TyOverby — DISMISSED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs (hunk 2 above)
Comment (TyOverby) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs]: I think that you've changed the BOM on this file.

Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs]: Github seems to be incorrectly showing a diff, when none exists. We have been ignoring it for all PRs, doing the same for this one.

Comment (khyperia) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs]: You have indeed changed (removed) the BOM.

image

Comment (khyperia) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs]: Sorry, got my diff backwards - you added the BOM (which is correct style for roslyn code, iirc)

Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs]: @TyOverby @khyperia - I just opened the file in VS to edit. @333fred Is it likely that our Operation unit test baseline updater is changing the BOM?

Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs]: I am merging in as the PR seems to be adding the correct BOM.

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs]: @mavasani probably, I was just using File.WriteAllText with no encoding specifier. I'll update it to use the existing encoding.

Comment (TyOverby) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs]: @333fred You might be interested in the patch that I made correcting BOM encoding in my autoformatter.

```diff
diff --git a/Program.cs b/Program.cs
index 4654046..059d0bc 100644
--- a/Program.cs
+++ b/Program.cs
@@ -8,6 +8,7 @@
 using static Microsoft.CodeAnalysis.Formatting.Formatter;
 using Microsoft.CodeAnalysis.Text;
 using System.Diagnostics;
+using System.Text;
 
 namespace CSharpFormatter
 {
@@ -86,6 +87,7 @@ static void ApplyChangesToDocument(string documentPath, IEnumerable<DiffRequest>
         {
             if (documentPath.EndsWith("vb")) { return; }
             SyntaxNode after;
+            Encoding encoding;
             using (var file = File.Open(documentPath, FileMode.Open))
             {
                 var sourceText = SourceText.From(file);
@@ -96,10 +98,12 @@ static void ApplyChangesToDocument(string documentPath, IEnumerable<DiffRequest>
                 // need to be able to adjust them.
                 requests = requests.OrderBy(r => r.lineStart);
                 var spans = requests.Select(req => SpanForLines(sourceText, req.lineStart, req.lineCount));
+
                 after = Format(tree.GetRoot(), spans, new AdhocWorkspace());
+                encoding = sourceText.Encoding;
             }
 
-            File.WriteAllText(documentPath, after.ToFullString());
+            File.WriteAllText(documentPath, after.ToFullString(), encoding);
         }
 
         static IEnumerable<string> ConsoleLines()

```

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs]: I would indeed, I was just figured out how to get the current file encoding.

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs]: @mavasani pushed an update to my fixer that uses the correct file encoding.

Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs]: Thanks Fred.

Comment (TyOverby) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs]: @333fred Could you run this fixer over all of roslyn in order to correct any other files that may have been mistakenly switched?

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs]: @TyOverby this fixer isn't designed for that, it scans for IOperation tests that need to have their baselines updated. Your formatter with the process all roslyn switch is much closer.

Comment (TyOverby) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs]: @333fred My formatter won't add the bom on files that don't have it.  It'll correctly write it back the way it was written.

**Pull request merged**